### PR TITLE
Improve framework generalization

### PR DESCRIPTION
**Pull request by kentjhall**
don't merge yet pls

### CODE REVIEW

**Comment by hmontero1205** [comment]
TODO:
- ~pylint xD~
- documentation
- ~ok i know ive called variables `CHUNGUS` before but since this is public facing, let's eventually rename it to `NAME` or something~

**Comment by kentjhall** [comment]
i felt the `CHUNGUS` was tasteful :((

File: common/grades.py [hunk 1]
```diff
@@ -167,13 +171,18 @@ def print_submission_grades(
...
+                continue
             for item_code, rubric_item in rubric_item_mappings.items():
                 if(rubric_code != "ALL"
                    and not item_code.startswith(rubric_code)
                    or not self.is_graded(f"{item_code}.1", name)):
                     continue
                 graded = True
+                if rubric_item.deduct_from:
+                    floor_pts = total_pts
+                    total_pts += rubric_item.deduct_from
```
Comment (hmontero1205) [inline common/grades.py:185]: I think you should add some comments here...

File: common/grades.py [hunk 2]
```diff
@@ -200,6 +209,9 @@ def print_submission_grades(
 
                     if item_comments:
                         all_comments.append(item_comments)
+                if rubric_item.deduct_from:
+                    ceiled = min(floor_pts + rubric_item.deduct_from, total_pts)
+                    total_pts = max(floor_pts, ceiled)
```
Comment (hmontero1205) [inline common/grades.py:221]: ...and here

File: common/grades.py (hunk 1 above)
Comment (hmontero1205) [inline common/grades.py:189]: I think you should add comments here...

File: common/grades.py [hunk 3]
```diff
@@ -208,8 +220,11 @@ def print_submission_grades(
             all_comments.insert(0, "(LATE)")
         concatted_comments = "; ".join(all_comments)
 
-        if rubric_code == "ALL" and self.is_late(name):
-            total_pts = round(total_pts * (1 - LATE_PENALTY), 2)
+        late_penalty = DEFAULT_LATE_PENALTY
+        if "late_penalty" in self.rubric:
+            late_penalty = self.rubric["late_penalty"]
+        if rubric_code == "ALL" and self.is_late(name) and late_penalty != 0:
```
Comment (hmontero1205) [inline common/grades.py]: `late_penalty == 0` is ok, I think we can omit this check

File: common/grades.py [hunk 4]
```diff
@@ -208,8 +220,11 @@ def print_submission_grades(
...
+            total_pts = round(total_pts * (1 - late_penalty), 2)
```
Comment (hmontero1205) [inline common/grades.py:234]: not asking you to do this, just leaving this as food for thought

perhaps we should call a generic `apply_late_penalty(total_pts)` and have the user plug in their own logic. I imagine not all classes want to roll with percent deduction

File: common/hw_base.py [hunk 5]
```diff
@@ -71,12 +72,19 @@ def create_rubric(self, rubric_file):
 
         rubric = {}
         for table_k, table_v in rubric_json.items():
+            if table_k == "late_penalty":
+                rubric[table_k] = table_v
+                continue
```
Comment (hmontero1205) [inline common/hw_base.py:77]: another note to self, perhaps we're starting to reach a point where keeping the rubric as a pure python dict is getting silly. we might need to replace this was a class or something

File: common/submissions.py [hunk 6]
```diff
@@ -46,9 +44,9 @@ def check_late(deadline_path, iso_timestamp):
                        f"and {secs} secs late")
     return True
 
-def checkout_to_team_master(
-        repo: git.Repo, hw_name: str, team: str) -> bool:
-    repo.git.checkout("master")  # Make sure we're already on master (skel).
+def checkout_to_team_branch(
+        repo: git.Repo, team_repo_id: str, team: str, branch_name: str) -> bool:
+    repo.git.checkout(branch_name)  # Make sure we're on the branch (skel).
```
Comment (hmontero1205) [inline common/submissions.py:53]: I'm not sure this is right. Sorry this is on me because I never explained how this worked to anyone and didn't write it down in plain english.

The idea is that the TA gets a copy of the HW skel code in their grading workspace. That means that in that repo, the `master` branch refers to the `master` branch of the skel repo (ie. the initial commit we gave to students).

Now let's say we want to check out to a team's specific branch (I hardcoded in team's `master`). I first checkout the workspace repo to its `master` (skel code, see line 49) and then checkout to the team's `master` (I called the branch `team`, see line 75. A better name wouldve been `team-master`). That is what I am deleting on line 68 just in case.

TBH my process was prolly overkill.. I wrote the initial logic for this in rush 1hr before our first grading meeting in Spring 2020 and was paranoid about checkout conflicts (if that's even a thing xD). I ported the bash logic to Python over the summer without thinking any deeper.

Anyways, given all of that, can you make sure this code is actually right?

lol as I am writing this sentence, I now realize I think your code is equivalent to mine if branch_name == master. if that's the case, i apologize for the long story, pls lmk xD

Comment (hmontero1205) [inline common/submissions.py:53]: Maybe we should reword the comment.. something to make clearer that we assume our version of `branch_name` is the skeleton code and the team's version of `branch_name` builds off of the skel.

File: common/submissions.py [hunk 7]
```diff
@@ -46,9 +44,9 @@ def check_late(deadline_path, iso_timestamp):
                        f"and {secs} secs late")
     return True
 
-def checkout_to_team_master(
-        repo: git.Repo, hw_name: str, team: str) -> bool:
-    repo.git.checkout("master")  # Make sure we're already on master (skel).
+def checkout_to_team_branch(
+        repo: git.Repo, team_repo_id: str, team: str, branch_name: str) -> bool:
```
Comment (hmontero1205) [inline common/submissions.py]: i think the function needs a docstring

File: common/submissions.py [hunk 8]
```diff
@@ -59,29 +57,33 @@ def checkout_to_team_master(
...
     except git.GitError:
         # This branch doesn't exist.
         pass
 
-    repo.git.checkout("-b", team, f"{team}/master")
+    repo.git.checkout("-b", team_branch, f"{team}/{branch_name}")
 
     return True
 
+def checkout_to_team_master(
```
Comment (hmontero1205) [inline common/submissions.py]: what's the point of this function? if it's for backwards compatibility, it still wont worth because of the extra param right?

File: common/submissions.py [hunk 9]
```diff
@@ -118,6 +120,45 @@ def checkout_to_tag_then_test(hw_instance):
...
+    """Decorator function that checks out submitter-branch_name before the test.
+
+    This is assumed to be in reference to a submission branch, which follows the
+    naming convention 'submitter-branch_name' (see checkout_to_team_branch()).
+    If the checkout fails, we open a shell for the grader to resolve the issue.
+    """
+
+    def function_wrapper(test_func):
+        def checkout_to_branch_then_test(hw_instance):
+            nonlocal branch_name
```
Comment (hmontero1205) [inline common/submissions.py:163]: (not to self) lol pylint somehow didnt see this and still threw unused-argument? this is so cursed lol

I know you based this off of my `tag` implementation, but I wonder if I did it in a cringe way.. i def dont have enough knowledge of decorators.

File: hw_setup.py [hunk 10]
```diff
@@ -73,83 +58,21 @@ def main():
...
     create_dir(root)
 
-    pygrader_dir = os.getcwd()
+    pygrader_dir = Path(__file__).resolve().parent
 
     os.chdir(root)
-    if args.hw in ('hw1', 'linux-list'):
-        if not s_path:
-            sys.exit("hw1 usage: ./hw_setup hw1 -s <submission zipfile path>")
-        if not args.submissions or (not os.path.isfile(s_path) or
```
Comment (hmontero1205) [inline hw_setup.py:87]: i uhhh might need this XD could we somehow kick this over to `common/submissions.py`?

File: hw_setup.py [hunk 11]
```diff
@@ -73,83 +58,21 @@ def main():
...
+    if os.path.isdir(args.hw):
+        _prompt_overwrite(args.hw, args.hw)
+    create_dir(args.hw)
+    os.chdir(args.hw)
 
-    else:
-        sys.exit(f"Unsupported assignment: {args.hw}")
+    setup_script = os.path.join(pygrader_dir, args.hw, 'setup')
+    if os.path.isfile(setup_script) and os.access(setup_script, os.X_OK):
+        os.system(setup_script)
```
Comment (hmontero1205) [inline hw_setup.py]: this is epic

File: rubric.json.in [hunk 12]
```diff
@@ -0,0 +1,3 @@
+{
+	"late_penalty": 0.2
```
Comment (hmontero1205) [inline rubric.json.in]: im not sure if this is a good template rubric

do you think `late_penalty: 0` and maybe defining rubric item A1 with 0 pts and a dummy msg would be better?

**Review by hmontero1205 — CHANGES_REQUESTED** [review]
Additional TODO:
- document deductive rubric

**Comment by hmontero1205** [comment]
image
@kentjhall imagine not having a big chungus reference somewhere in the codebase

File: hw_setup.py (hunk 10 above)
Comment (kentjhall) [reply hw_setup.py:87]: yeah so it should go in `hw1/setup` when we rebase the cs4118 version eventually. ideally it won't be in this repo

File: common/submissions.py (hunk 8 above)
Comment (kentjhall) [reply common/submissions.py]: yeah but i left it because cs4118 only ever checks out to master, so its j kinda nicer looking. now that i think about it tho, it might make more sense to do `branch_name = "master"` default param instead, so then cs4118 would j replace every call to `checkout_to_team_master` with `checkout_to_team_branch`, i think i'll do that

File: common/grades.py (hunk 4 above)
Comment (kentjhall) [reply common/grades.py:234]: i agree with this for the future, i was thinking we should at least add an additional option for flat point deductions (especially doable since now it's specified per-assignment) but this is probably a better idea

File: common/hw_base.py (hunk 5 above)
Comment (kentjhall) [reply common/hw_base.py:77]: yeah, this was originally the route I was going but then realized that I don't understand Python typing well enough xD

File: rubric.json.in (hunk 12 above)
Comment (kentjhall) [reply rubric.json.in]: sure, i guess my idea was to have this as an example value to show that it should be a decimal between [0, 1], but doesn't rly matter (though I think 1.0 makes more sense as a default value then 0.0, if there's no deduction then is it even really "late"?)

File: common/submissions.py (hunk 9 above)
Comment (kentjhall) [reply common/submissions.py:163]: yeah me neither lol, i tried to google around and understand but i still don't get what's going on there

File: grade.py [hunk 13]
```diff
@@ -157,6 +158,8 @@ def prompt_grade(self, rubric_item: RubricItem):
         """Prompts the TA for pts/comments"""
         for i, (pts, desc) in enumerate(rubric_item.subitems, 1):
             subitem_code = f"{rubric_item.code}.{i}"
+            if self.grades.is_graded(subitem_code):
```
Comment (hmontero1205) [inline grade.py]: I'm confused, isn't this what the default mode is anyways? pygrader skips things that are already graded if you didnt run with --regrade
